From af11ab06fd50da809ce937b9e571ed3fc9a1e819 Mon Sep 17 00:00:00 2001
From: Aravind Mosam <amosam@qti.qualcomm.com>
Date: Sun, 6 Oct 2024 23:24:17 +0530
Subject: [PATCH] feeds: Add support for fetching feeds from mirror

In case of failure to fetch from one url, try for other urls
in feeds.conf
Add the support via option -b

Signed-off-by: Aravind Mosam <amosam@qti.qualcomm.com>
---
 include/toplevel.mk |  2 +-
 scripts/feeds       | 12 +++++++++---
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/include/toplevel.mk b/include/toplevel.mk
index 658d30e414f6..594beb68ea32 100644
--- a/include/toplevel.mk
+++ b/include/toplevel.mk
@@ -263,7 +263,7 @@ endif
 
 # update all feeds, re-create index files, install symlinks
 package/symlinks:
-	./scripts/feeds update -a
+	./scripts/feeds update -a -b
 	./scripts/feeds install -a -f
 
 # re-create index files, install symlinks
diff --git a/scripts/feeds b/scripts/feeds
index 7cbe07f58e7c..0cdc00d433ba 100755
--- a/scripts/feeds
+++ b/scripts/feeds
@@ -788,12 +788,13 @@ sub uninstall {
 	return 0;
 }
 
-sub update_feed($$$$$$)
+sub update_feed($$$$$$$)
 {
 	my $type=shift;
 	my $name=shift;
 	my $src=shift;
 	my $force_update=shift;
+	my $fallback=shift;
 	my $rebase_update=shift;
 	my $stash_update=shift;
 	my $force_relocate=update_location( $name, "@$src" );
@@ -817,6 +818,10 @@ sub update_feed($$$$$$)
 				warn "failed, ignore.\n";
 				next;
 			}
+			if ($fallback) {
+				warn "Failed to  fetch '$name' from '$feedsrc', trying next \n";
+				next;
+			}
 			last;
 		}
 		$failed = 0;
@@ -836,7 +841,7 @@ sub update {
 	$ENV{SCAN_COOKIE} = $$;
 	$ENV{OPENWRT_VERBOSE} = 's';
 
-	getopts('ahifrs', \%opts);
+	getopts('ahifbrs', \%opts);
 	%argv_feeds = map { $_ => 1 } @ARGV;
 
 	if ($opts{h}) {
@@ -858,7 +863,7 @@ sub update {
 		my ($type, $name, $src) = @$feed;
 		next unless $#ARGV == -1 or $opts{a} or $argv_feeds{$name};
 		if (not $opts{i}) {
-			update_feed($type, $name, $src, $opts{f}, $opts{r}, $opts{s}) == 0 or $failed=1;
+			update_feed($type, $name, $src, $opts{f}, $opts{b}, $opts{r}, $opts{s}) == 0 or $failed=1;
 		}
 		push @index_feeds, $name;
 	}
@@ -927,6 +932,7 @@ Commands:
 	    -s :           Update by rebase and autostash. (git only. Useful if local commits and uncommited changes exist)
 	    -i :           Recreate the index only. No feed update from repository is performed.
 	    -f :           Force updating feeds even if there are changed, uncommitted files.
+	    -b :           Support fetching from mirrors
 
 	clean:             Remove downloaded/generated files.
 
-- 
2.34.1

